modules/core/04-channel/keeper/upgrade.go

```diff
@@ -0,0 +1,1015 @@
+	}
+
+	if connection.GetState() != int32(connectiontypes.OPEN) {
+		return errorsmod.Wrapf(
+			connectiontypes.ErrInvalidConnectionState,
+			"connection state is not OPEN (got %s)", connectiontypes.State(connection.GetState()).String(),
+		)
+	}
+
+	getVersions := connection.GetVersions()
+	if len(getVersions) != 1 {
+		return errorsmod.Wrapf(
+			connectiontypes.ErrInvalidVersion,
+			"single version must be negotiated on connection before opening channel, got: %v",
+			getVersions,
+		)
+	}
+
+	if !connectiontypes.VerifySupportedFeature(getVersions[0], proposedUpgrade.Ordering.String()) {
+		return errorsmod.Wrapf(
+			connectiontypes.ErrInvalidVersion,
+			"connection version %s does not support channel ordering: %s",
+			getVersions[0], proposedUpgrade.Ordering.String(),
+		)
+	}
+
+	return nil
+}
+
+// extractUpgradeFields returns the upgrade fields from the provided channel.
+func extractUpgradeFields(channel types.Channel) types.UpgradeFields {
+	return types.UpgradeFields{
+		Ordering:       channel.Ordering,
+		ConnectionHops: channel.ConnectionHops,
+		Version:        channel.Version,
+	}
+}
+
+// MustAbortUpgrade will restore the channel state to its pre-upgrade state so that upgrade is aborted.
+// Any unnecessary state is deleted and an error receipt is written.
+// This function is expected to always succeed, a panic will occur if an error occurs.
+func (k Keeper) MustAbortUpgrade(ctx sdk.Context, portID, channelID string, err error) {
+	if err := k.abortUpgrade(ctx, portID, channelID, err); err != nil {
+		panic(err)
+	}
+}
+
+// abortUpgrade will restore the channel state to its pre-upgrade state so that upgrade is aborted.
+// Any unnecessary state is delete and an error receipt is written.
+func (k Keeper) abortUpgrade(ctx sdk.Context, portID, channelID string, err error) error {
+	if err == nil {
+		return errorsmod.Wrap(types.ErrInvalidUpgradeError, "cannot abort upgrade handshake with nil error")
+	}
+
+	channel, found := k.GetChannel(ctx, portID, channelID)
+	if !found {
+		return errorsmod.Wrapf(types.ErrChannelNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	// in the case of application callbacks, the error may not be an upgrade error.
+	// in this case we need to construct one in order to write the error receipt.
+	upgradeError, ok := err.(*types.UpgradeError)
+	if !ok {
+		upgradeError = types.NewUpgradeError(channel.UpgradeSequence, err)
+	}
+
+	// the channel upgrade sequence has already been updated in ChannelUpgradeTry, so we can pass
+	// its updated value.
+	k.restoreChannel(ctx, portID, channelID, channel.UpgradeSequence, channel, upgradeError)
+	return nil
+}
+
+// restoreChannel will restore the channel state to its pre-upgrade state so that upgrade is aborted.
+func (k Keeper) restoreChannel(ctx sdk.Context, portID, channelID string, upgradeSequence uint64, channel types.Channel, err *types.UpgradeError) types.Channel {
```

Comment:
Do we ever use a sequence other than channel sequence here?

Reply:
I think this func is flagged to removed in another PR

Reply:
I think when the channel sequence is told to fast forward by the counterparty (counterparty error receipt is greater than our sequence)?